src/main/java/org/reactome/addlinks/db/ReferenceObjectCache.java

```diff
@@ -106,183 +106,217 @@ private static synchronized void buildReferenceCaches(String className, Map<Stri
 					e.printStackTrace();
+					throw new Error(e);
 				}
+
+			}
+		}
+		catch (InvalidAttributeException e)
+		{
+			logger.error("Object {} does not have an identifier attribute, error: {}", referenceObject, e.getMessage());
+			e.printStackTrace();
+		}
+		catch (Exception e)
+		{
+			e.printStackTrace();
+			throw new Error(e);
+		}
+	}
+
+	/**
+	 * Adds an object to a cache that is indexed by ReferenceDatabase.
+	 * @param cacheByRefDB - the cache to add to.
+	 * @param referenceObject - the object to add.
+	 */
+	private static void addObjectToRefDBCache(Map<String, List<GKInstance>> cacheByRefDB, GKInstance referenceObject)
+	{
+		String refDBID;
+		try
+		{
+			refDBID = String.valueOf(((GKInstance) referenceObject.getAttributeValue(ReactomeJavaConstants.referenceDatabase)).getDBID());
+			if (! cacheByRefDB.containsKey(refDBID))
+			{
+				List<GKInstance> byRefDb = Collections.synchronizedList( new LinkedList<GKInstance>() );
+				byRefDb.add(referenceObject);
+				cacheByRefDB.put(refDBID, byRefDb);
+			}
+			else
+			{
+				cacheByRefDB.get(refDBID).add(referenceObject);
+			}
+		}
+		catch (InvalidAttributeException e)
+		{
+			logger.error("Could not get the \"{}\" attribute for {} because it not valid for this object.", ReactomeJavaConstants.referenceDatabase, referenceObject);
+		}
+		catch (Exception e)
+		{
+			logger.error("Could not get the \"{}\" attribute for {}. Reason: {}", ReactomeJavaConstants.identifier, referenceObject, e.getMessage());
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * Adds an object to a cache that is indexed by species.
+	 * @param className - the className of the object.
+	 * @param cacheBySpecies - the cache to add the object to.
+	 * @param referenceObject - the object to add.
+	 */
+	private static void addObjectToSpeciesCache(String className, Map<String, List<GKInstance>> cacheBySpecies, GKInstance referenceObject)
+	{
+		// Sets do not allow duplicates.
+		Set<String> allSpecies = null;
+		//ReferenceRNASequence objects don't always have Species info, for some reason, so we need to get that from the associated ReferenceGeneProduct
+		if (className.equals(ReactomeJavaConstants.ReferenceRNASequence))
+		{
+			Object species = null;
+			try
+			{
+				species = referenceObject.getAttributeValue(ReactomeJavaConstants.species);
+			}
+			catch (InvalidAttributeException e)
+			{
+				logger.error("Could not get the \"{}\" attribute for {} because it not valid for this object.", ReactomeJavaConstants.species, referenceObject);
+			}
+			catch (Exception e)
+			{
+				logger.error("Could not get the \"{}\" attribute for {}. Reason: {}", ReactomeJavaConstants.species, referenceObject, e.getMessage());
 			}
-			// DB_ID Cache
-			cacheByID.put(String.valueOf(referenceObject.getDBID()), referenceObject);
 			
-			// Identifier cache
-			if ( !className.equals(ReactomeJavaConstants.Reaction) )
+			if (null == species)
```

Comment:
Perhaps a comment for what will specifically happen in cases where the `species` object is null? To me it looks like it iterates through `referenceTranscript` referrals to try and get any species, but some explanation would help